File: content/snippets/typescript/exercises/advanced/dijkstra_algorithm.ts
type Edge = [number,number]
const g: Record<number,Edge[]> = {0:[[1,10],[2,3]],1:[[3,2]],2:[[1,1],[3,8]],3:[]}
const INF = 1e9
function dijkstra(src:number){ const dist = new Map<number,number>(); for(const k in g) dist.set(Number(k), INF); dist.set(src,0);
  const pq: [number,number][] = [[0,src]]
  while(pq.length){ pq.sort((a,b)=>a[0]-b[0]); const [d,u]=pq.shift()!; if(d!==dist.get(u)) continue; for(const [v,w] of g[u]) if((dist.get(v)??INF) > d + w){ dist.set(v,d+w); pq.push([d+w,v]); } }
  return Object.fromEntries(dist)
}
console.log(dijkstra(0))
export {}

File: content/snippets/python/exercises/advanced/dijkstra_algorithm.py
import heapq

g = {0:[(1,10),(2,3)], 1:[(3,2)], 2:[(1,1),(3,8)], 3:[]}
INF=10**9

def dijkstra(src):
    dist = {v:INF for v in g}
    dist[src]=0; pq=[(0,src)]
    while pq:
        d,u = heapq.heappop(pq)
        if d!=dist[u]: continue
        for v,w in g[u]:
            if dist[v]>d+w:
                dist[v]=d+w; heapq.heappush(pq,(dist[v],v))
    return dist

print(dijkstra(0))
